rust-stream/src/main.rs

```diff
@@ -0,0 +1,413 @@
+            timings.dot[i] = dot;
+            last_sum = sum;
+        }
+        (timings, last_sum)
+    }
+
+    fn run_triad(&mut self, n: usize) -> Duration {
+        timed(|| for _ in 0..n { self.triad(); })
+    }
+
+    fn run_nstream(&mut self, n: usize) -> Vec<Duration> {
+        (0..n).map(|_| timed(|| self.nstream())).collect::<Vec<_>>()
+    }
+}
+
+trait ArrayType: Real + NumAssign + Signed + Default {}
+
+impl<T: Real + NumAssign + Signed + Default> ArrayType for T {}
+
+// single threaded version
+impl<T: ArrayType> RustStream<T, PlainFor> for StreamData<T> {
+    fn init_arrays(&mut self, init: (T, T, T)) {
+        self.a.fill(init.0);
+        self.b.fill(init.1);
+        self.c.fill(init.2);
+    }
+
+    fn copy(&mut self) {
+        for i in 0..self.size {
+            self.c[i] = self.a[i];
+        }
+    }
+
+    fn mul(&mut self) {
+        for i in 0..self.size {
+            self.b[i] = self.scalar * self.c[i];
+        }
+    }
+
+    fn add(&mut self) {
+        for i in 0..self.size {
+            self.c[i] = self.a[i] + self.b[i];
+        }
+    }
+
+    fn triad(&mut self) {
+        for i in 0..self.size {
+            self.a[i] = self.b[i] + self.scalar * self.c[i];
+        }
+    }
+
+    fn nstream(&mut self) {
+        for i in 0..self.size {
+            self.a[i] += self.b[i] * self.scalar * self.c[i];
+        }
+    }
+
+    fn dot(&mut self) -> T {
+        let mut sum: T = T::default();
+        for i in 0..self.size {
+            sum += self.a[i] * self.b[i];
+        }
+        sum
+    }
+}
+
+//  Rayon version, it should be semantically equal to the single threaded version
+impl<T: ArrayType + Sync + Send + Sum> RustStream<T, RayonPar> for StreamData<T> {
+    fn init_arrays(&mut self, init: (T, T, T)) {
+        self.a.fill(init.0);
+        self.b.fill(init.1);
+        self.c.fill(init.2);
+    }
+
+    fn copy(&mut self) {
+        let a = &self.a;
+        self.c.par_iter_mut().enumerate().for_each(|(i, c)| *c = a[i])
+    }
+
+    fn mul(&mut self) {
+        let c = &self.c;
+        let scalar = &self.scalar;
+        self.b.par_iter_mut().enumerate().for_each(|(i, b)| *b = *scalar * c[i])
+    }
+
+    fn add(&mut self) {
+        let a = &self.a;
+        let b = &self.b;
+        self.c.par_iter_mut().enumerate().for_each(|(i, c)| *c = a[i] + b[i])
+    }
+
+    fn triad(&mut self) {
+        let scalar = &self.scalar;
+        let b = &self.b;
+        let c = &self.c;
+        self.a.par_iter_mut().enumerate().for_each(|(i, a)| *a = b[i] + *scalar * c[i])
+    }
+
+    fn nstream(&mut self) {
+        let scalar = &self.scalar;
+        let b = &self.b;
+        let c = &self.c;
+        self.a.par_iter_mut().enumerate().for_each(|(i, a)| *a += b[i] + *scalar * c[i])
+    }
+
+    fn dot(&mut self) -> T {
+        let a = &self.a;
+        let b = &self.b;
+        (0..self.size).into_par_iter().fold(|| T::default(), |acc, i| acc + a[i] * b[i]).sum::<T>()
```

Comment:
You can write this a little more idiomatically as:
```rust
self.a.par_iter().zip(&self.b).map(|(&a, &b)| a * b).sum()
```

Reply:
As Matt says, sequential floating point operations perform quite
badly in standard Rust as there is no "fast math" option to ignore the associativity
constraints. This leads to summation loops not vectorising as the LLVM autovectoriser
will not break the constraint.

(see https://stackoverflow.com/questions/30863510/how-do-i-compile-with-ffast-math for example).

You can write less idiomatic code that sums chunks using the `chunks_exact(..)` iterator
which will vectorise, but it is a shame that you have to. I am working on some code transformation
tools that may help in this case as part of our **extendr** R extension project.

I would generally avoid using SIMD crates and intrinsics unless you really have to.